### PR TITLE
boards/Board.mk: use genromfs to make romfs image

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
With this patch, user can use RCSRCS or RCRAW specify which files need put into romfs.

### CODE REVIEW

File: boards/Board.mk [hunk 1]
```diff
@@ -37,6 +37,27 @@
...
+RCOBJS = $(RCSRCS:%=%$(OBJEXT))
+
+$(RCOBJS): %$(OBJEXT): %
+	$(call PREPROCESS, $<, $@)
+	$(Q) mkdir -p $(dir $(ETCDIR)$(DELIM)$<)
+	$(Q) cp -r $@ $(ETCDIR)$(DELIM)$<
+
+$(ETCSRC): $(shell find $(RCRAWS) -type l) $(RCOBJS)
+	$(foreach raw,$(RCRAWS), \
+		$(shell cp --parents -rfLp $(raw) $(ETCDIR)))
```
Comment (yamt) [inline boards/Board.mk:55]: i suspect --parents is not portable.

usually -R is preferred over -r these days. do you need -r?

**Comment by yamt** [comment]
> With this patch, user can use RCSRCS or RCRAW specify which files need put into romfs.

i guess some documentation is necessary.

**Comment by xiaoxiang781216** [comment]
> what will use this ?

Now NuttX use mkromfs.sh+Kconfig to generate the romfs content which is very difficult to modify. With RCSRCS and RCRAWS, user can add any files to romfs:
1.Files in RCSRCS is processed by C preprocessor which is useful for the shell script
2.Files in RCRAWS is directly put into romfs which is sutiable for some binary content

File: boards/Board.mk [hunk 2]
```diff
@@ -37,6 +37,27 @@
 
 -include $(TOPDIR)/Make.defs
 
+ifneq ($(RCSRCS)$(RCRAWS),)
+ETCDIR := etctmp
+ETCSRC := $(ETCDIR:%=%.c)
+
+CSRCS += $(ETCSRC)
+
+RCOBJS = $(RCSRCS:%=%$(OBJEXT))
```
Comment (yamt) [inline boards/Board.mk:46]: is OBJEXT appropriate for this?

File: boards/Board.mk [hunk 3]
```diff
@@ -37,6 +37,27 @@
...
+
+$(RCOBJS): %$(OBJEXT): %
+	$(call PREPROCESS, $<, $@)
+	$(Q) mkdir -p $(dir $(ETCDIR)$(DELIM)$<)
+	$(Q) cp -r $@ $(ETCDIR)$(DELIM)$<
+
+$(ETCSRC): $(shell find $(RCRAWS) -type l) $(RCOBJS)
```
Comment (yamt) [inline boards/Board.mk:53]: why do you list symbolic links here?

File: boards/Board.mk [hunk 4]
```diff
@@ -138,6 +164,8 @@ endif
 
 clean:
 	$(call DELFILE, libboard$(LIBEXT))
+	$(call DELFILE, $(ETCSRC))
+	$(call DELDIR, $(ETCDIR))
```
Comment (yamt) [inline boards/Board.mk:168]: clean RCOBJS here?

**Comment by xiaoxiang781216** [comment]
> is this an alternative to NSH_CUSTOMROMFS?
> 

RCSRCS/RCRAWS isn't an alternative which is designed to work with NSH_ROMFSETC/NSH_ARCHROMFS/NSH_CUSTOMROMFS. In the old workflow, if user enable NSH_ROMFSETC, he/she need prepare nsh_romfsimg.h by self or utilize mkromfs.sh(but this approach is only suitable for demo purpose). In the new workflow, user just need:
1.Enable NSH_ROMFSETC
2.Append script files into RCSRCS
2.Append binary files to RCRAWS
The build system will automatically generate romfs.img and convert to header file for you.

> how is a user supposed to specify RCSRCS and RCRAWS?

Here is a demo boards/arm/chip/board/src/Makefile
```
RCSRCS := rcS

VPATH += :etc/init.d
DEPPATH += --dep-path etc/init.d

include $(TOPDIR)/boards/Board.mk
```

And boards/arm/chip/board/src/etc/init.d/rcS
```
#include <nuttx/config.h>

#ifdef CONFIG_NSH_PROC_MOUNTPOINT
mount -t procfs CONFIG_NSH_PROC_MOUNTPOINT
#endif

#ifdef CONFIG_FS_TMPFS
mount -t tmpfs CONFIG_LIBC_TMPDIR
#endif

#ifndef CONFIG_NSH_DISABLE_UNAME
uname -a > /dev/syslog
#endif
```

File: boards/Board.mk (hunk 1 above)
Comment (xiaoxiang781216) [reply boards/Board.mk:55]: Fixed.

File: boards/Board.mk (hunk 2 above)
Comment (xiaoxiang781216) [reply boards/Board.mk:46]: Fixed.

File: boards/Board.mk (hunk 3 above)
Comment (xiaoxiang781216) [reply boards/Board.mk:53]: Fixed.

File: boards/Board.mk (hunk 4 above)
Comment (xiaoxiang781216) [reply boards/Board.mk:168]: RCOBJ is put into ETCDIR, so remove $(ETCDIR) is enough.

**Comment by xiaoxiang781216** [comment]
> > With this patch, user can use RCSRCS or RCRAW specify which files need put into romfs.
> 
> i guess some documentation is necessary.

@anchao will provide a document.

**Comment by patacongo** [comment]
It has always been a policy not to add any new tools to the build process.  If tools are added, the must be fully verified on all build platforms and documented in the top level README.txt file.  We must clearly advise people how to set up the environment on every supported platform.

**Comment by xiaoxiang781216** [comment]
> @xiaoxiang781216 -
> 
> > RCSRCS/RCRAWS isn't an alternative which is designed to work with NSH_ROMFSETC/NSH_ARCHROMFS/NSH_CUSTOMROMFS.
> 
> I am confused by the statement. " isn't an alternative"
> 
> To be clear: Will the old way still work Yes or No?
> 
> If the answer is No. This change should not come in, without deprecating the old way and announcing its End Of Life at a future time.

Of course, the old method still can be used without any change, we seldom provide a solution stop the old way. If the compatiblity can't achieve, we will provide the patch to convert all usage to the new method.

**Comment by xiaoxiang781216** [comment]
> It has always been a policy not to add any new tools to the build process. If tools are added, the must be fully verified on all build platforms and documented in the top level README.txt file. We must clearly advise people how to set up the environment on every supported platform.

Yes, we will convert one config to utilize the new method which will:
1.Ensure the new method work in all supported platform
2.Demostrate how to use the new method to simplify the work

**Comment by davids5** [comment]
@xiaoxiang781216 - that is all good. I think the wording was just not clear:  "it is an alternative"

**Comment by patacongo** [comment]
> Yes, we will convert one config to utilize the new method which will:
> 1.Ensure the new method work in all supported platform
> 2.Demostrate how to use the new method to simplify the work

A page on the Confluence Wiki would probably be even better.   A config is good for testing, but it is not good for documenting or communicating.

**Comment by xiaoxiang781216** [comment]
> > Yes, we will convert one config to utilize the new method which will:
> > 1.Ensure the new method work in all supported platform
> > 2.Demostrate how to use the new method to simplify the work
> 
> A page on the Confluence Wiki would probably be even better. A config is good for testing, but it is not good for documenting or communicating.

Ok, @anchao will find a time to document the new method, but the workflow is very simple now. User just need add script or file to RCSRCS/RCRAWS and then the build system will do all the thing automatically. No need to understand genromfs, xxd and nsh_romfsimg.h anymore.

@yamt The first patch address all your issue.
@patacongo @davids5 the second patch demo how simple and flexible with the new method
Of course the old method still work as before.

**Comment by patacongo** [comment]
> Ok, @anchao will find some time to document the new method, but the workflow is very simple now. User just need add script or file to RCSRCS/RCRAWS and then the build system will do all the thing automatically. No need to understand genromfs, xxd and nsh_romfsimg.h anymore.

I can help with the writing so don't let the language differences be a barrier to communicating well.

**Comment by yamt** [comment]
i guess the README needs to be updated. esp. the "Updating the ROMFS File System" section.
otherwise lgtm.

**Comment by xiaoxiang781216** [comment]
> i guess the README needs to be updated. esp. the "Updating the ROMFS File System" section.
> otherwise lgtm.

Done.

**Pull request merged**